sdk/cosmosdb/cosmos/src/utils/batch.ts

```diff
@@ -210,6 +212,53 @@ export function decorateOperation(
   return operation as Operation;
 }
 
+/**
+ * Splits a batch into array of batches based on cumulative size of its operations by making sure
+ * cumulative size of an individual batch is not larger than {@link Constants.DefaultMaxBulkRequestBodySizeInBytes}.
+ * If a single operation itself is larger than {@link Constants.DefaultMaxBulkRequestBodySizeInBytes}, that
+ * operation would be moved into a batch containing only that operation.
+ * @param originalBatch - A batch of operations needed to be checked.
+ * @returns
+ * @hidden
+ */
+export function splitBatchBasedOnBodySize(originalBatch: Batch): Batch[] {
+  if (originalBatch?.operations === undefined && originalBatch.operations.length < 1) return [];
+  let currentBatchSize = calculateObjectSizeInBytes(originalBatch.operations[0]);
+  let currentBatch: Batch = {
+    ...originalBatch,
+    operations: [originalBatch.operations[0]],
+    indexes: [originalBatch.indexes[0]],
+  };
+  const processedBatches: Batch[] = [];
+  processedBatches.push(currentBatch);
+
+  for (let index = 1; index < originalBatch.operations.length; index++) {
+    const operation = originalBatch.operations[index];
+    const currentOpSize = calculateObjectSizeInBytes(operation);
+    if (currentBatchSize + currentOpSize > Constants.DefaultMaxBulkRequestBodySizeInBytes) {
+      currentBatch = {
+        ...originalBatch,
+        operations: [],
+        indexes: [],
+      };
+      processedBatches.push(currentBatch);
+      currentBatchSize = 0;
+    }
+    currentBatch.operations.push(operation);
+    currentBatch.indexes.push(originalBatch.indexes[index]);
+    currentBatchSize += currentOpSize;
+  }
+  return processedBatches;
+}
+
+/**
+ * Calculates size of an JSON object in bytes with utf-8 encoding.
+ * @hidden
+ */
+export function calculateObjectSizeInBytes(obj: unknown): number {
+  return new TextEncoder().encode(bodyFromData(obj as any)).length;
```

Comment:
There's something off here. The length of a string in JS isn't measured in bytes, it's measured in 16-bit code units, so if you want to know the byte-length of a string, you have to multiply by two. However, you aren't trying to compute the length of the body as a string, but of the body object itself. For this, if `bodyFromData` returns a typed array view (ArrayBufferView), you can use the `byteLength` property.

If any non-ascii characters are included in the body, the string-length of the encoded body will not match the byte length of the buffer exactly.

Reply:
Nevermind... I have it backwards. `TextEncoder.encode` is producing a typed array, not the other way around. This operation might be a little expensive if the body is large.